Sources/Testing/Traits/Trait.swift

```diff
@@ -41,6 +41,68 @@ public protocol Trait: Sendable {
+
+  /// The type of the custom test executor for this trait.
+  ///
+  /// The default type is `Never`.
+  associatedtype CustomTestExecutor: CustomTestExecuting = Never
+
+  /// The custom test executor for this trait, if any.
+  ///
+  /// If this trait's type conforms to ``CustomTestExecuting``, the default
+  /// value of this property is `self` and this trait will be used to customize
+  /// test execution. This is the most straightforward way to implement a trait
+  /// which customizes the execution of tests.
+  ///
+  /// However, if the value of this property is an instance of another type
+  /// conforming to ``CustomTestExecuting``, that instance will be used to
+  /// perform custom test execution instead.  Otherwise, the default value of
+  /// this property is `nil` (with the default type `Never?`), meaning that
+  /// custom test execution will not be performed for tests this trait is
+  /// applied to.
+  var customTestExecutor: CustomTestExecutor? { get }
+}
+
+/// A protocol that allows customizing the execution of a test function (and
+/// each of its cases) or a test suite by performing custom code before or after
+/// it runs.
+public protocol CustomTestExecuting: Sendable {
+  /// Execute a function for the specified test and/or test case.
+  ///
+  /// - Parameters:
+  ///   - function: The function to perform. If `test` represents a test suite,
+  ///     this function encapsulates running all the tests in that suite. If
+  ///     `test` represents a test function, this function is the body of that
+  ///     test function (including all cases if it is parameterized.)
+  ///   - test: The test under which `function` is being performed.
+  ///   - testCase: The test case, if any, under which `function` is being
+  ///     performed. This is `nil` when invoked on a suite.
+  ///
+  /// - Throws: Whatever is thrown by `function`, or an error preventing
+  ///   execution from running correctly.
+  ///
+  /// This function is called for each ``Trait`` on a test suite or test
+  /// function which has a non-`nil` value for ``Trait/customTestExecutor-1dwpt``.
+  /// It allows additional work to be performed before or after the test runs.
+  ///
+  /// This function is invoked once for the test its associated trait is applied
+  /// to, and then once for each test case in that test, if applicable. If a
+  /// test is skipped, this function is not invoked for that test or its cases.
+  ///
+  /// Issues recorded by this function are associated with `test`.
+  func execute(_ function: @Sendable () async throws -> Void, for test: Test, testCase: Test.Case?) async throws
+}
+
+extension Trait where CustomTestExecutor == Self {
+  public var customTestExecutor: CustomTestExecutor? {
+    self
+  }
+}
+
+extension Never: CustomTestExecuting {
+  public func execute(_ function: @Sendable () async throws -> Void, for test: Test, testCase: Test.Case?) async throws {
+    fatalError("Unreachable codepath: Never cannot be instantiated.")
```

Comment:
Do we even need this `fatalError()` given this is unreachable and the function doesn't need to return anything? 🤔